### PR TITLE
refactor(core): css-keyframes feature extraction

### PR DESCRIPTION
**Pull request by idoros**
## fix
- [x] unresolved imported keyframes have no diagnostics - [see reproduction](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%281%21%27%2540st-import%2520%255Bkeyframes%257B3%257D%255D%2520from%2520%255C%27.2%255C%2740.root%2520%280%2520%2520%2520%2520animation-name%253A%252034%29%27%7E2%21%27%27%29%7Eselected%211%27%29*.st.css0%255Cn1%27%252Findex*2%252Fother*3unknown4%253B0%250143210*_)
- [x] change current nesting diagnostics that only reports under `@st-scope` to report under any unknown nesting with the exception of `@media` and stylesheet root
- [x] don't report `@keyframes` override under different `@media` rules  ([see demo](https://jsbin.com/gefepofanu/1/edit?html,css,output)) - **keep report on import override or any other collision**.
- [x] fix global keyframes with no name from registering a symbol

## refactor
- [x] move related code into the feature (`KeyframesSymbol`, `reservedKeyFrames`)
- [x] processor - move code to `analyzeAtRule`
- [x] transformer - split keyframes resolve/transform/export into the general flow
  - `transformResolve` (new hook) - move `resolveKeyframes` to feature
  - `transformAtRuleNode` (new hook) - move to feature
  - `transformDeclaration` (new hook) - move `animation/animation-name` namespace to feature
  - `transformJSExports` (new hook) - move to feature
- [x] change `st-import` feature to register keyframes to `st-symbol`
  - [x] make the registration plugable from `css-keyframes`
- [x] align `redeclare` keyframes symbol with other symbol diagnostics through `st-symbol`

## deprecate
- [x] `meta.keyframes`
- [x] `meta.mappedKeyframes`

## tests
- [x] move tests to `features/css-keyframes.spec`
- validate feature tests completeness
  - [x] align tests structure
  - [x] add missing tests
    - check `keyframes` symbol collection
    - check local override of imported symbol
    - check diagnostics report for redeclare keyframes
    - js-mixin with `st-global`, unknown and known keyframes mixed-in

## move to backlog
 - [ ] "re-organizing/merge" feature tests template, because currently it is splitted into `meta(analyze), transform, and diagnostics` which causes a lot of duplications - having a single test-kit tool to check collected analyze data, symbols, and diagnostics + transformations and exports would help to combine specs
 - [ ] refactor `css-class` resolve / exports code into feature
 - [ ] `st-mixin` from JS symbols are not part of the mixed-in validation or transformation - for example `animation-name` in mixin cannot resolve `@keyframes` that are defined in the same mixin ([see reproduction](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%283%21%27%2540st-import%25207%2520from%25201.%252F7180.root5-st-7in%253A%252078%290%27%7E%252F7.js%21%27module.exports%2520%253D%2520funcA%257B%257D5return5*1%2540keyframes%2520a2%28%2961.x2%28B**1expect24%253Dentry__a16*421a16%29B%298%29%27%29%7Eselected%213%27%29*990%255Cn1%255C%272%255C%21%25203%27%252Findex.st.css41animaA-name5%2520%28B6%252CB*7mix8%253B09%2520%2520AtionB0*%2501BA9876543210*_)) 
- [ ] keyframe escaping bug #1930

### CODE REVIEW

File: packages/core/src/features/st-symbol.ts [hunk 1]
```diff
@@ -1,8 +1,9 @@
 import { FeatureContext, createFeature } from './feature';
```
[the remark this answers is not in the document]
Comment (idoros) [reply packages/core/src/features/st-symbol.ts:1]: they are not conflicting with other symbols, so they cannot be on `mappedSymbols`.

filtered out non `main` namespace symbols from deprecated `mappedSymbols`

File: packages/core/src/features/css-keyframes.ts [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            ast: atRule,
+            global,
+        });
+    },
+    transformResolve({ context }) {
+        const symbols = STSymbol.getAllByType(context.meta, `keyframes`);
+        const resolved: Record<string, KeyframesResolve> = {};
+        for (const [name, symbol] of Object.entries(symbols)) {
+            const res = resolveKeyframes(context.meta, symbol, context.resolver);
+            if (res) {
```
[the remark this answers is not in the document]
Comment (idoros) [reply packages/core/src/features/css-keyframes.ts:144]: good catch!

File: packages/cli/src/base-generator.ts [hunk 3]
```diff
@@ -111,7 +112,7 @@ export function reExportsAllSymbols(filePath: string, generator: Generator): ReE
         acc[varName] = `--${rootExport}__${varName.slice(2)}`;
         return acc;
     }, {});
-    const keyframes = Object.keys(meta.mappedKeyframes).reduce<Record<string, string>>(
+    const keyframes = Object.keys(CSSKeyframes.getAll(meta)).reduce<Record<string, string>>(
```
[the remark this answers is not in the document]
Comment (idoros) [reply packages/cli/src/base-generator.ts:115]: I prefer not to open public API unless it is specifically required

File: packages/core/src/stylable-resolver.ts [hunk 4]
```diff
@@ -129,7 +130,7 @@ export class StylableResolver {
             const symbol =
                 !name || subtype === `mappedSymbols`
                     ? meta.getSymbol(name || meta.root)!
-                    : meta.mappedKeyframes[name];
+                    : STSymbol.get(meta, name, `keyframes`)!;
```
[the remark this answers is not in the document]
Comment (idoros) [reply packages/core/src/stylable-resolver.ts]: I cleaned the code a bit to have a single call to STSymbol, I prefer not to expose `keyframes` symbol for now, maybe we can change the interface before publicly exposing it.

**Review by barak007 — APPROVED** [review]

File: packages/core/test/features/css-keyframes.spec.ts [hunk 5]
```diff
@@ -0,0 +1,724 @@
...
+            });
+
+            expect(
+                CSSKeyframes.getKeyframesStatements(meta),
+                `CSSKeyframes.getKeyframesStatements(meta)`
+            ).to.containSubset([meta.ast.nodes[0], meta.ast.nodes[1]]);
+
+            // deprecation
+            ignoreDeprecationWarn(() => {
+                expect(meta.keyframes.length).to.eql(2);
```
Comment (tzachbon) [inline packages/core/test/features/css-keyframes.spec.ts:44]: Use the same check as above

File: packages/core/test/features/css-keyframes.spec.ts [hunk 6]
```diff
@@ -0,0 +1,724 @@
...
+        });
+        describe(`st-import`, () => {
+            it('should namespace imported animation and animation name', () => {
+                const result = generateStylableRoot({
+                    entry: `/entry.st.css`,
+                    files: {
+                        '/entry.st.css': {
+                            namespace: 'entry',
+                            content: `
+                                :import {
```
Comment (tzachbon) [inline packages/core/test/features/css-keyframes.spec.ts:225]: Refactor to `@st-import`

File: packages/core/test/features/css-keyframes.spec.ts [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+                expectTransformDiagnostics(config, []);
+            });
+            it('should not conflict with other named parts', () => {
+                const result = generateStylableRoot({
+                    entry: `/entry.st.css`,
+                    files: {
+                        '/entry.st.css': {
+                            namespace: 'entry',
+                            content: `
+                                :import {
```
Comment (tzachbon) [inline packages/core/test/features/css-keyframes.spec.ts:328]: Refactor to `@st-import`

File: packages/core/test/features/css-keyframes.spec.ts [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+                    const cssExports = generateStylableExports(config);
+
+                    expect(cssExports.keyframes).to.eql({
+                        globalName: 'globalName',
+                    });
+                });
+            });
+        });
+        describe(`escape`, () => {
+            it.skip(`should escape invlid inputs`, () => {
```
Comment (tzachbon) [inline packages/core/test/features/css-keyframes.spec.ts:509]: fix typo `invalid`

**Review by tzachbon — APPROVED** [review]
Looks great, left some comments

**Review by tomrav — APPROVED** [review]

File: packages/core/src/features/css-keyframes.ts [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+        }
+        //
+        let global: boolean | undefined;
+        const globalName = globalValue(name);
+        if (globalName !== undefined) {
+            name = globalName;
+            global = true;
+        }
+        if (name === '') {
+            context.diagnostics.warn(atRule, diagnostics.MISSING_KEYFRAMES_NAME_INSIDE_GLOBAL());
```
[the remark this answers is not in the document]
Comment (idoros) [reply packages/core/src/features/css-keyframes.ts:127]: fixed

**Pull request merged**